tuned_lens/scripts/train_loop.py

```diff
@@ -16,246 +18,362 @@
     shift_preds,
     send_to_device,
 )
+from tuned_lens.scripts.ingredients import (
+    local_rank,
+    world_size,
+    Model,
+    Data,
+    Sharding,
+)
 import torch as th
-import torch.distributed as dist
-
-
-def train_loop(
-    args: Namespace,
-    model: th.nn.Module,
-    data: Dataset,
-    lens: TunedLens,
-    nats_to_bpb: float,
-):
-    """Trains a TunedLens model against a transformer on a dataset.
-
-    Args:
-        args: The command-line arguments see __main__.py train subcommand.
-        model: The transformer model to train.
-        data: The dataset to train on.
-        lens: The TunedLens model to train.
-        nats_to_bpb: The ratio of nats to bits per byte for the dataset.
-    """
-    lens_size = sum(p.numel() * p.element_size() for p in lens.parameters())
-    print(f"Tuned lens memory usage: {lens_size / 2 ** 20:.2f} MB per GPU")
-
-    if args.constant:
-        for probe in lens:
-            probe.weight.requires_grad_(False)
-
-    local_rank = dist.get_rank() if dist.is_initialized() else 0
-    world_size = dist.get_world_size() if dist.is_initialized() else 1
-    if world_size > 1:
-        ddp_lens = DDP(lens, device_ids=[local_rank], find_unused_parameters=True)
-    else:
-        ddp_lens = lens
-
-    dl = DataLoader(
-        data.shuffle(seed=args.seed),  # type: ignore[arg-type]
-        batch_size=args.per_gpu_batch_size,
-    )
-    if args.wandb and local_rank == 0:
+from dataclasses import dataclass
+
+
+class LossChoice(enum.Enum):
```

Comment:
Sigh, I have a general policy of using `typing.Literal` instead of `enum.Enum` since `enum.Enum` is quite verbose, but it's ultimately your call.

Reply:
Simple parsing does not support this https://github.com/lebrice/SimpleParsing/issues/2.